MazerPlatformer/Character.cs

```diff
@@ -103,30 +110,30 @@ public void MoveInDirection(CharacterDirection direction, GameTime dt)
                     X += MoveByStep();
                     break;
                 default:
-                    throw new ArgumentOutOfRangeException(nameof(direction), direction, null);
+                    return new InvalidDirectionFailure(direction);
             }
-            SetCharacterDirection(direction);
+            return SetCharacterDirection(direction);
         }
 
-        private void SetState(CharacterStates state)
+        private Either<IFailure, Unit> SetState(CharacterStates state)
         {
             CurrentState = state;
-            OnStateChanged?.Invoke(state);
+            return Ensure( () => OnStateChanged?.Invoke(state));
         }
 
-        private void SetCharacterDirection(CharacterDirection direction)
-        {
-            CurrentDirection = direction;
-
-            SetAnimationDirection(direction);
-            OnDirectionChanged?.Invoke(direction);
-            SetState(CharacterStates.Moving);
+        private Either<IFailure, Unit> SetCharacterDirection(CharacterDirection direction) 
+            => SetAnimationDirection(direction)
+                .Iter(unit => 
+            {
+                CurrentDirection = direction;
+                OnDirectionChanged?.Invoke(direction);
+                SetState(CharacterStates.Moving);
 
-            Animation.Idle = false;
-            CanMove = true;
-        }
+                Animation.Idle = false;
+                CanMove = true;
+            });
 
-        private void SetAnimationDirection(CharacterDirection direction)
+        private Either<IFailure, Unit> SetAnimationDirection(CharacterDirection direction)
```

Comment:
In this Scenario, instead of setting the CurrentAnimationDirection on the underlying Animation object, you pass in the animation object, set it and then pass it back. This fixes the problem of underlying state changing in the underlying class that the function is operating on